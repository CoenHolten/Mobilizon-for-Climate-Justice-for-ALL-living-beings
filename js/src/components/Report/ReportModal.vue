<template>
        <div class="modal-card">
            <header class="modal-card-head" v-if="title">
                <p class="modal-card-title">{{ title }}</p>
            </header>

            <section
                    class="modal-card-body is-flex"
                    :class="{ 'is-titleless': !title }">
                <div class="media">
                    <div
                            class="media-left">
                        <b-icon
                                icon="alert"
                                type="is-warning"
                                size="is-large"/>
                    </div>
                    <div class="media-content">
                        <p>The report will be sent to the moderators of your instance.
                            You can explain why you report this content below.</p>

                        <div class="control">
                            <b-input
                                    v-model="content"
                                    type="textarea"
                                    @keyup.enter="confirm"
                                    placeholder="Additional comments"
                            />
                        </div>

                        <p v-if="outsideDomain">
                            The content came from another server. Transfer an anonymous copy of the reportÂ ?
                        </p>

                        <div class="control" v-if="outsideDomain">
                            <b-switch v-model="forward">Transfer to {{ outsideDomain }}</b-switch>
                        </div>
                    </div>
                </div>
            </section>

            <footer class="modal-card-foot">
                <button
                        class="button"
                        ref="cancelButton"
                        @click="close">
                    {{ cancelText }}
                </button>
                <button
                        class="button is-primary"
                        ref="confirmButton"
                        @click="confirm">
                    {{ confirmText }}
                </button>
            </footer>
        </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';
import { removeElement } from 'buefy/src/utils/helpers';

@Component({
  mounted() {
    this.$data.isActive = true;
  },
})
export default class ReportModal extends Vue {
  @Prop({ type: Function, default: () => {} }) onConfirm;
  @Prop({ type: String }) title;
  @Prop({ type: String, default: '' }) outsideDomain;
  @Prop({ type: String, default: 'Cancel' }) cancelText;
  @Prop({ type: String, default: 'Send the report' }) confirmText;

  isActive: boolean = false;
  content: string = '';
  forward: boolean = false;

  confirm() {
    this.onConfirm(this.content, this.forward);
    this.close();
  }

    /**
     * Close the Dialog.
     */
  close() {
    this.isActive = false;
    this.$emit('close');
  }
}
</script>
<style lang="scss">
    .modal-card .modal-card-foot {
        justify-content: flex-end;
    }
</style>