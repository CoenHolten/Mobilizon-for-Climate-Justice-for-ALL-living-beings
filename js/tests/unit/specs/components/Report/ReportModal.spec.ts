import { config, mount } from "@vue/test-utils";
import ReportModal from "@/components/Report/ReportModal.vue";
import { vi, beforeEach, describe, it, expect } from "vitest";

import Oruga from "@oruga-ui/oruga-next";

config.global.plugins.push(Oruga);

const propsData = {
  onConfirm: vi.fn(),
};

const generateWrapper = (customPropsData: Record<string, unknown> = {}) => {
  return mount(ReportModal, {
    props: {
      ...propsData,
      ...customPropsData,
    },
  });
};

beforeEach(() => {
  vi.resetAllMocks();
});

describe("ReportModal", () => {
  it("renders report modal with basic informations and submits it", async () => {
    const wrapper = generateWrapper();

    expect(wrapper.find("header").exists()).toBe(false);

    expect(wrapper.find("section").text()).not.toContain(
      "The content came from another server. Transfer an anonymous copy of the report?"
    );

    expect(wrapper.find("footer button:first-child").text()).toBe("Cancel");

    const submit = wrapper.find("footer button.o-btn--primary");

    expect(submit.text()).toBe("Send the report");

    const textarea = wrapper.find("textarea");
    textarea.setValue("some comment with my report");

    submit.trigger("click");

    await wrapper.vm.$nextTick();
    expect(wrapper.emitted().close).toBeTruthy();

    expect(wrapper.vm.$props.onConfirm).toHaveBeenCalledTimes(1);
    expect(wrapper.vm.$props.onConfirm).toHaveBeenCalledWith(
      "some comment with my report",
      false
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders report modal and shows an inline comment if it's provided", async () => {
    const wrapper = generateWrapper({
      comment: {
        actor: { preferredUsername: "author", name: "I am the comment author" },
        text: "this is my <b>comment</b> that will be reported",
      },
    });

    const commentContainer = wrapper.find("article.media");
    expect(commentContainer.find("strong").text()).toContain(
      "I am the comment author"
    );
    expect(commentContainer.find("small").text()).toContain("author");
    expect(commentContainer.find("p").html()).toContain(
      "this is my <b>comment</b> that will be reported"
    );
  });

  it("renders report modal with with a remote content", async () => {
    const wrapper = generateWrapper({ outsideDomain: "somewhere.else" });

    expect(wrapper.find(".control p").text()).toContain(
      "The content came from another server. Transfer an anonymous copy of the report?"
    );

    const submit = wrapper.find("footer button.o-btn--primary");
    submit.trigger("click");
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.$props.onConfirm).toHaveBeenCalledTimes(1);
    expect(wrapper.vm.$props.onConfirm).toHaveBeenCalledWith("", false);
  });

  it("renders report modal with with a remote content and accept to forward", async () => {
    const wrapper = generateWrapper({ outsideDomain: "somewhere.else" });

    expect(wrapper.find(".control p").text()).toContain(
      "The content came from another server. Transfer an anonymous copy of the report?"
    );

    const switchButton = wrapper.find('input[type="checkbox"]');
    switchButton.setValue(true);

    const submit = wrapper.find("footer button.o-btn--primary");
    submit.trigger("click");
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.$props.onConfirm).toHaveBeenCalledTimes(1);
    expect(wrapper.vm.$props.onConfirm).toHaveBeenCalledWith("", true);
  });

  it("renders report modal custom title and buttons", async () => {
    const wrapper = generateWrapper({
      title: "want to report something?",
      cancelText: "nah",
      confirmText: "report!",
    });

    expect(wrapper.find("header h2").text()).toBe("want to report something?");

    expect(wrapper.find("footer button:first-child").text()).toBe("nah");

    expect(wrapper.find("footer button.o-btn--primary").text()).toBe("report!");
  });
});
