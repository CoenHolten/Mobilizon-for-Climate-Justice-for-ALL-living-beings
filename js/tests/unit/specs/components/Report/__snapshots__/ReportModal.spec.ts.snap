// Vitest Snapshot v1

exports[`ReportModal > renders report modal with basic informations and submits it 1`] = `
"<div class=\\"p-2\\" data-v-8c6db6e4=\\"\\">
  <!--v-if-->
  <section data-v-8c6db6e4=\\"\\">
    <div class=\\"flex gap-1 flex-row mb-3\\" data-v-8c6db6e4=\\"\\"><span class=\\"o-icon o-icon--warning hidden md:block flex-1\\" data-v-8c6db6e4=\\"\\"><i class=\\"mdi mdi-alert 48\\"></i></span>
      <p data-v-8c6db6e4=\\"\\">The report will be sent to the moderators of your instance. You can explain why you report this content below.</p>
    </div>
    <div class=\\"\\" data-v-8c6db6e4=\\"\\">
      <!--v-if-->
      <div class=\\"o-field o-field--filled\\" data-v-8c6db6e4=\\"\\"><label for=\\"additonal-comments\\" class=\\"o-field__label\\">Additional comments</label>
        <div class=\\"o-ctrl-input\\" data-v-8c6db6e4=\\"\\"><textarea id=\\"additonal-comments\\" class=\\"o-input o-input__textarea\\"></textarea>
          <!--v-if-->
          <!--v-if-->
          <!--v-if-->
        </div>
        <!--v-if-->
      </div>
      <!--v-if-->
    </div>
  </section>
  <footer class=\\"flex gap-2 py-3\\" data-v-8c6db6e4=\\"\\"><button type=\\"button\\" class=\\"o-btn\\" data-v-8c6db6e4=\\"\\"><span class=\\"o-btn__wrapper\\"><!--v-if--><span class=\\"o-btn__label\\">Cancel</span>
      <!--v-if--></span>
    </button><button type=\\"button\\" class=\\"o-btn o-btn--primary\\" data-v-8c6db6e4=\\"\\"><span class=\\"o-btn__wrapper\\"><!--v-if--><span class=\\"o-btn__label\\">Send the report</span>
      <!--v-if--></span>
    </button></footer>
</div>"
`;
